src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs

```diff
@@ -126,86 +106,60 @@ private TextFormattingRunProperties Format
+
+                using var _1 = SegmentedListPool.GetPooledList<TagSpan<InlineHintDataTag>>(out var dataTagSpans);
+                _underlyingTagger.AddTags(spans, dataTagSpans);
+
+                // Presize so we can add the elements below without continually resizing.
+                adornmentTagSpans.Capacity += dataTagSpans.Count;
+
+                using var _2 = PooledHashSet<int>.GetInstance(out var seenPositions);
 
-                var selectedSpans = new List<TagSpan<IntraTextAdornmentTag>>();
-                for (var i = 0; i < _cache.Count; i++)
+                var format = this.Format;
+                foreach (var dataTagSpan in dataTagSpans)
                 {
-                    var tagSpans = _cache[i].mappingTagSpan.Span.GetSpans(snapshot);
-                    if (tagSpans.Count == 1)
-                    {
-                        var tagSpan = tagSpans[0];
-                        if (spans.IntersectsWith(tagSpan))
-                        {
-                            if (_cache[i].tagSpan is not { } hintTagSpan)
-                            {
-                                var hintUITag = InlineHintsTag.Create(
-                                        _cache[i].mappingTagSpan.Tag.Hint, Format, _textView, tagSpan, _taggerProvider, _formatMap, classify);
-
-                                hintTagSpan = new TagSpan<IntraTextAdornmentTag>(tagSpan, hintUITag);
-                                _cache[i] = (_cache[i].mappingTagSpan, hintTagSpan);
-                            }
-
-                            selectedSpans.Add(hintTagSpan);
-                        }
-                    }
+                    if (seenPositions.Add(dataTagSpan.Span.Start))
+                        adornmentTagSpans.Add(GetOrCreateAdornmentTagsSpan(dataTagSpan, colorHints, format));
                 }
-
-                return selectedSpans;
             }
             catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, ErrorSeverity.General))
             {
                 throw ExceptionUtilities.Unreachable();
             }
         }
 
-        public void Dispose()
+        private TagSpan<IntraTextAdornmentTag> GetOrCreateAdornmentTagsSpan(
+            TagSpan<InlineHintDataTag> dataTagSpan, bool classify, TextFormattingRunProperties format)
         {
-            _tagAggregator.BatchedTagsChanged -= TagAggregator_BatchedTagsChanged;
-            _tagAggregator.Dispose();
-            _formatMap.ClassificationFormatMappingChanged -= OnClassificationFormatMappingChanged;
+            // If we've never computed the adornment info, or options have changed, then compute and cache the new information.
+            var cachedTagInformation = (CachedAdornmentTagSpan?)dataTagSpan.Tag.AdditionalData;
+            if (cachedTagInformation is null || cachedTagInformation.Classified != classify || cachedTagInformation.Format != format)
```

Comment:
no need for a complicated cache.  we have the data tag.  it can just hold onto this data for us.